File: src/com/colacelli/ircbot/plugins/rejoinonkick/RejoinOnKickPlugin.java
package com.colacelli.ircbot.plugins.rejoinonkick;

import com.colacelli.ircbot.IRCBot;
import com.colacelli.ircbot.Plugin;
import com.colacelli.irclib.actors.Channel;
import com.colacelli.irclib.actors.User;
import com.colacelli.irclib.connection.Connection;
import com.colacelli.irclib.connection.Rawable;
import com.colacelli.irclib.connection.listeners.Listener;
import com.colacelli.irclib.connection.listeners.OnKickListener;
import com.colacelli.irclib.connection.listeners.OnRawCodeListener;

import java.util.ArrayList;
import java.util.Timer;
import java.util.TimerTask;

public class RejoinOnKickPlugin implements Plugin {
    private ArrayList<Listener> listeners;

    public RejoinOnKickPlugin() {
        listeners = new ArrayList<>();

        listeners.add((OnKickListener) (connection, user, channel) -> {
            if (user.getNick().equals(connection.getUser().getNick())) connection.join(new Channel(channel.getName()));
        });

        listeners.add(new OnRawCodeListener() {
            @Override
            public int rawCode() {
                return Rawable.RawCode.JOIN_BANNED.getCode();
            }

            @Override
            public void onRawCode(Connection connection, String message, int rawCode, String... args) {
                Timer timer = new Timer();
                timer.schedule(new TimerTask() {
                    @Override
                    public void run() {
                        connection.join(new Channel(args[3]));
                    }
                }, 5000);
            }
        });
    }

    @Override
    public String name() {
        return "REJOIN_ON_KICK";
    }

    @Override
    public void onLoad(IRCBot bot) {
        listeners.forEach((listener) -> bot.addListener(listener));
    }

    @Override
    public void onUnload(IRCBot bot) {
        listeners.forEach((listener) -> bot.removeListener(listener));
    }
}

File: src/main/kotlin/com/colacelli/ircbot/plugins/rejoinonkick/RejoinOnKickPlugin.kt
package com.colacelli.ircbot.plugins.rejoinonkick

import com.colacelli.ircbot.IRCBot
import com.colacelli.ircbot.base.Plugin
import com.colacelli.irclib.actors.Channel
import com.colacelli.irclib.actors.User
import com.colacelli.irclib.connection.Connection
import com.colacelli.irclib.connection.Rawable
import com.colacelli.irclib.connection.listeners.Listener
import com.colacelli.irclib.connection.listeners.OnKickListener
import com.colacelli.irclib.connection.listeners.OnRawCodeListener
import com.colacelli.irclib.messages.PrivateMessage
import java.util.*

class RejoinOnKickPlugin : Plugin {
    private val listeners: ArrayList<Listener> = ArrayList()

    init {
        listeners.add(object : OnKickListener {
            override fun onKick(connection: Connection, user: User, channel: Channel) {
                return connection.join(channel)
            }
        })

        // TODO: Add other blocking entrance modes like +i, +l, etc.
        listeners.add(object : OnRawCodeListener {
            override fun rawCode(): Int {
                return Rawable.RawCode.JOIN_BANNED.code
            }

            override fun onRawCode(connection: Connection, message: String, rawCode: Int, args: List<String>) {
                val timer = Timer()
                timer.schedule(object : TimerTask() {
                    override fun run() {
                        val channel = Channel(args[3])

                        connection.send(PrivateMessage(
                                "unban ${channel.name}",
                                connection.user,
                                User("ChanServ")
                        ))
                        connection.join(channel)
                    }
                }, 5000)
            }
        })
    }

    override var name = "rejoin_on_kick"

    override fun onLoad(bot: IRCBot) {
        listeners.forEach {
            bot.addListener(it)
        }
    }

    override fun onUnload(bot: IRCBot) {
        listeners.forEach {
            bot.removeListener(it)
        }
    }
}